<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <title>Differential Expression Analysis Report</title>
    <link rel="stylesheet" href="media/styles.css" />
    <%
      parse.name <- function (n) (paste(unlist(strsplit(n, "_vs_")), collapse = " vs "))
      navlink <- function (link, text) {
        cat(paste0('<li><a href="',link,'">',text,'</a></li>'))
      }
      navmenu <- function (id, text, submenu) {
        cat('<li>')
        cat('<a href="#',id,'" data-toggle="collapse" aria-expanded="false" class="dropdown-toggle">',text,'</a>', sep="")
        cat('<ul class="collapse list-unstyled" id="',id,'">', sep="")
        submenu()
        cat('</ul></li>')
      }
      subsection <- function (id, title, content, shown=FALSE) {
        cat('<h2 id="',id,'" class="h4"><a href="#',id,'_container" data-toggle="collapse">',title,'</a></h2>', sep="")
        cat('<div id="',id,'_container" class="row collapse', ifelse(shown, ' show', ''),'"><div class="col">', sep="")
        content()
        cat('</div></div><div class="line"></div>')
      }
      subsection.one.image <- function (id, title, figure, message, caption, alt=caption) {
        subsection(id, title, function() {
          path <- gsub(PROJECT.PATH$main,".",figure)
          cat('<p class="text-justify">',message,'</p>', sep="")
          cat('<div class="text-center"><a href="',path,'" data-fancybox data-caption="',caption,'">',
            '<img src="',path,'" alt="',alt,'" class="img-fluid img-thumbnail"></a></div>', sep="")
        });
      }
      .last.gallery <- 0
      gallery_id <- function () {
        .last.gallery <<- .last.gallery + 1
        return (paste0('gallery', .last.gallery))
      }
      subsection.multi.image <- function (id, title, figures, message, captions, groups=NULL, alts=captions, fancycaptions=captions) {
        subsection(id, title, function() {
          path <- gsub(PROJECT.PATH$main,".",figures)
          cat('<p class="text-justify">',message,'</p>', sep="")
          if (is.null(groups)) {
            groups <- rep("", length(figures));
          }
          tapply(1:length(path), groups, function (ii) {
            gid <- gallery_id()
            if (groups[ii[1]] != "") {
              cat("<h6>",groups[ii[1]],"</h6>", sep="")
            }
            cat('<div class="row">')
            for (i in ii) {
              cat('<div class="col text-center"><figure class="figure">',
                  '<a href="',path[i],'" data-fancybox="',gid,'" data-caption="',fancycaptions[i],'">',
                  '<img src="',path[i],'" alt="',alts[i],'" class="img-fluid img-thumbnail"></a>',
                  '<figcaption class="figure-caption">',captions[i],'</figcaption></figure></div>',sep="")
            }
            cat('</div>')
          })
        });
      }
      subsection.one.gallery <- function (id, title, figures, message, caption, alt=caption) {
        caption <- setNames(rep_len(caption, length(figures)), path)
        alt     <- setNames(rep_len(alt, length(figures)), path)
        subsection.multi.image(id, title, figures, message, caption, NULL, alt)
      }
    %>
  </head>
  <body>
    <div class="wrapper">
      <!-- Sidebar Holder -->
      <nav id="sidebar">
        <div class="sidebar-header">
          <h3>Analysis Report</h3>
        </div>

        <ul id="sidebar-menu" class="list-unstyled components">
          <%
          navmenu('summary-menu', 'Summary', function () {
            navlink('#summary_summary', 'Summary')
            navlink('#summary_input', 'Input options')
            navlink('#summary_analysis_filtered', paste0("Filtered ",trans.level,"s"))
            navlink('#summary_analysis_nde', paste0("DE ",trans.level,"s"))
            if (run.log)  navlink('#summary_analysis_log', 'Run log')
          })
          if (exists("fig.raw") || exists("fig.unorm") || exists("fig.norm") || exists("fig.other")) {
            navmenu('figures-menu', 'Figures', function () {
              if (!is.null(fig.raw$mds)) navlink('#figures_mds', 'MDS')
              if (!is.null(fig.raw$biodetection)) navlink('#figures_biodetection', 'Biotype detection')
              if (!is.null(fig.raw$countsbio)) navlink('#figures_countsbio', 'Biotype counts')
              if (!is.null(fig.raw$saturation)) navlink('#figures_saturation', 'Biotype saturation<')
              if (!is.null(fig.raw$readnoise)) navlink('#figures_readnoise', 'Reads noise')
              if (!is.null(fig.raw$correl)) navlink('#figures_correl', 'Correlation plots')
              if (!is.null(fig.raw$pairwise)) navlink('#figures_pairwise', 'Pairwise scatterplots')
              if (!is.null(fig.raw$rnacomp)) navlink('#figures_rnacomp', 'RNA composition')
              if (!is.null(fig.unorm$boxplot) || !is.null(fig.norm$boxplot)) navlink('#figures_boxplot', 'Boxplots')
              if (!is.null(fig.unorm$gcbias) || !is.null(fig.norm$gcbias)) navlink('#figures_gcbias', 'GC-bias')
              if (!is.null(fig.unorm$lengthbias) || !is.null(fig.norm$lengthbias)) navlink('#figures_lengthbias', 'Length bias')
              if (!is.null(fig.unorm$meandiff) || !is.null(fig.norm$meandiff)) navlink('#figures_meandiff', 'Mean-difference')
              if (!is.null(fig.unorm$meanvar) || !is.null(fig.norm$meanvar)) navlink('#figures_meanvar', 'Mean-variance')
              if (!is.null(fig.other$filtered)) navlink('#figures_filtered', paste0("Filtered ",trans.level,"s"))
              if (!is.null(fig.stat$volcano)) navlink('#figures_volcano', 'Volcano plots')
              if (!is.null(fig.stat$deheatmap)) navlink('#figures_heatmap', 'DEG heatmaps')
              if (!is.null(fig.stat$biodist))  navlink('#figures_biodist', 'DEG biotypes')
              if (!is.null(fig.venn$venn)) navlink('#figures_venn', 'Venn diagrams')
            });
          }
          navmenu('results-menu', 'Results', function () {
            nn <- names(html)
            counter <- 1
            for (n in nn) {
              navlink(paste0('#table_container_',counter), parse.name(n))
              counter <- counter + 1
            }
          })
          navlink('#references', 'References')
          %>
        </ul>
      </nav>
      <div id="content">
        <nav class="navbar navbar-expand-lg navbar-light bg-light">
          <div class="container-fluid">
            <button type="button" id="sidebarCollapse" class="navbar-btn">
              <span></span>
              <span></span>
              <span></span>
            </button>
          </div>
        </nav>

        <div class="row">
          <div class="col">
            <h1 id="summary" class="section-header"><a href="#summary_container" data-toggle="collapse">Summary</a></h1>
          </div>
        </div>
        <div id="summary_container" class="row collapse show m-2">
          <div class="col">
            <h2 id="summary_summary" class="h4"><a href="#summary_summary_container" data-toggle="collapse">Analysis summary</a></h2>
            <div id="summary_summary_container" style="text-align:justify;" class="collapse show">
              <p><strong>Summary: </strong>
              <%
              if (from.raw) {
                cat("The raw ",file.type," files, one for each RNA-Seq sample, were summarized to ")
                if (trans.level=="gene") {
                  if (count.type=="exon") cat("an exon ")
                  else if (count.type=="gene") cat("a gene ")
                  else if (count.type=="utr") cat("a 3'UTR ")
                } else if (trans.level=="transcript") {
                  if (count.type=="exon") cat("an exon ")
                  else if (count.type=="utr") cat("a 3'UTR ")
                } else if (trans.level=="exon") {
                  if (count.type=="exon") cat("an exon ")
                }
                cat("read counts table, ")
                if (file.type=="bam" || file.type=="sam")
                  cat("using the Bioconductor package GenomicRanges. ")
                else if (file.type=="bed")
                  cat("using the Bioconductor packages rtracklayer and GenomicRanges. ")
                cat("In the final read counts table, each row represented ")
                if (count.type=="exon") cat("one exon, ") else if (count.type=="gene") cat("one gene, ")
                cat("each column one RNA-Seq sample and each cell, the corresponding read counts associated with each row and column.")
              }
              if (count.type=="exon") {
                if (!is.null(exon.filters)) {
                  cat("The exon read counts were filtered for artifacts that could affect the subsequent normalization and statistical ")
                  cat("testing procedures as follows: ")
                  msg.exon.filters.min.active.exons <- NULL
                  if (!is.null(exon.filters$min.active.exons)) {
                    msg.exon.filters.min.active.exons <- paste0(
                      "if an annotated ",trans.level," had up to ",exon.filters$min.active.exons$exons.per.gene," exons, read presence was required in at ",
                      "least ",exon.filters$min.active.exons$min.exons," of the exons. Otherwise, if an annotated ",trans.level," had more than ",
                      exon.filters$min.active.exons$exons.per.gene," exons, then read presence was required in at least ",exon.filters$min.active.exons$frac,
                      "x&lceil;E&rceil; exons, where &lceil;<sup>.</sup>&rceil; is the <em>ceiling</em> mathematical function. The application ",
                      "of this filter resulted in the exclusion of ",length(exon.filter.result$min.active.exons)," ",trans.level,"s from further analysis. ")
                  }
                  cat(msg.exon.filters.min.active.exons,sep=", ")
                  cat("The total number of ",trans.level,"s excluded due to the application of exon filters was ",sum(as.numeric(sapply(exon.filter.result,length))),". ",sep="")
                }
                cat("The final read counts for each ",trans.level," model were calculated as the sums of their exon reads, creating a ",trans.level," counts ")
                cat("table where each row corresponded to an Ensembl ",trans.level," model and each column corresponded to an RNA-Seq sample. ")
              }
              cat("The ",trans.level," counts table was normalized for inherent systematic or experimental biases (e.g. sequencing depth, ",trans.level," length, ")
              cat("GC content bias etc.) using the Bioconductor package ",report.messages$norm[[normalization]]," after removing ",trans.level,"s that had ", sep="")
              cat("zero counts over all the RNA-Seq samples (",length(the.zeros)," ",trans.level,"s). The output of the normalization algorithm ",sep="")
              cat("was a table with normalized counts, which can be used for differential expression analysis with statistical algorithms ")
              cat("developed specifically for count data. ")
              if (!is.null(gene.filters)) {
                cat("Prior to the statistical testing procedure, the ",trans.level," read counts were filtered for possible artifacts that could affect ")
                cat("the subsequent statistical testing procedures. Genes/transcripts presenting any of the following were excluded from further analysis: ")
                latin.numbered <- c("i)","ii)","iii)","iv)","v)","vi)","vii)","viii)","ix)","x)","xi)","xii)","xiii)","xiv)","xv)")
                counter <- 0
                msg.gene.filters.length <- msg.gene.filters.expression.median <- msg.gene.filters.expression.mean <-
                msg.gene.filters.expression.quantile <- msg.gene.filters.expression.known <- msg.gene.filters.expression.custom <-
                msg.gene.filters.biotype <- msg.gene.filters.avg.reads <- msg.gene.filters.presence <- NULL
                if (!is.null(gene.filters$length)) {
                  counter <- counter + 1
                  msg.gene.filters.length <- paste0(latin.numbered[counter]," ",trans.level,"s with length less than ",gene.filters$length$length,
                                                    " (",length(gene.filter.result$length)," genes)")
                }
                if (!is.null(gene.filters$avg.reads)) {
                  counter <- counter + 1
                  msg.gene.filters.avg.reads <- paste0(
                    latin.numbered[counter]," ",trans.level,"s whose average reads per ",gene.filters$avg.reads$average.per.bp," bp was less than the ",
                    100*gene.filters$avg.reads$quantile,"<sup>th</sup> quantile of the total normalized distribution of average reads per ",
                    gene.filters$avg.reads$average.per.bp,"bp (",length(gene.filter.result$average.per.bp)," ",trans.level,"s with cutoff value ",
                    round(gene.filter.cutoff$avg.reads,digits=5)," average reads per ",gene.filters$avg.reads$average.per.bp," bp)")
                }
                if (!is.null(gene.filters$expression)) {
                  if (!is.null(gene.filters$expression$median) && gene.filters$expression$median) {
                    counter <- counter + 1
                    msg.gene.filters.expression.median <- paste0(
                      latin.numbered[counter]," ",trans.level,"s with read counts below the median read counts of the total normalized count distribution ",
                      "(",length(gene.filter.result$expression$median)," ",trans.level,"s with cutoff value ",gene.filter.cutoff$expression$median,
                      " normalized read counts)")
                  }
                  if (!is.null(gene.filters$expression$mean) && gene.filters$expression$mean) {
                    counter <- counter + 1
                    msg.gene.filters.expression.mean <- paste0(
                      latin.numbered[counter]," ",trans.level,"s with read counts below the mean read counts of the total normalized counts distribution ",
                      "(",length(gene.filter.result$expression$mean)," ",trans.level,"s with cutoff value ",gene.filter.cutoff$expression$mean,
                      " normalized read counts)")
                  }
                  if (!is.null(gene.filters$expression$quantile) && !is.na(gene.filters$expression$quantile)) {
                    counter <- counter + 1
                    msg.gene.filters.expression.quantile <- paste0(
                      latin.numbered[counter]," ",trans.level,"s with read counts below the ",100*gene.filters$expression$quantile,"<sup>th</sup> ",
                      "quantile of the normalized counts distribution (",length(gene.filter.result$expression$quantile)," ",trans.level,"s with cutoff value ",
                      gene.filter.cutoff$expression$quantile," normalized read counts)")
                  }
                  if (!is.null(gene.filters$expression$known) && !is.na(gene.filters$expression$known)) {
                    counter <- counter + 1
                    msg.gene.filters.expression.known <- paste0(
                      latin.numbered[counter]," ",trans.level,"s with read counts below the 90<sup>th</sup> quantile of the counts of the following ",
                      trans.level,"s, known to not being expressed from the related literature: ",paste(gene.filters$expression$known,collapse=", "),
                      "(",length(gene.filter.result$expression$known)," ",trans.level,"s with cutoff value",gene.filter.cutoff$expression$known,
                      " normalized read counts)")
                  }
                  if (!is.null(gene.filters$expression$custom) && !is.na(gene.filters$expression$custom)) {
                    counter <- counter + 1
                    msg.gene.filters.expression.custom <- paste0(
                      latin.numbered[counter]," genes not passing the user defined filter provided to the metaseqr pipeline ",
                      "(",length(gene.filter.result$expression$custom)," ",trans.level,"s with cutoff value",gene.filter.cutoff$expression$custom,")")
                  }
                }
                if (!is.null(gene.filters$biotype)) {
                  counter <- counter + 1
                  msg.gene.filters.biotype <- paste0(
                    latin.numbered[counter]," ",trans.level,"s whose biotype matched the following: ",
                    paste(names(gene.filters$biotype)[which(unlist(gene.filters$biotype))],collapse=", "),
                    " (",length(gene.filter.result$biotype)," ",trans.level,"s)"
                  )
                }
                if (!is.null(gene.filters$presence)) {
                  counter <- counter + 1
                  msg.gene.filters.presence <- paste0(
                    latin.numbered[counter]," ",trans.level,"s where ",gene.filters$presence$frac*100,"% of samples were not above ",
                    gene.filters$presence$min.count," counts (",length(gene.filter.result$presence)," ",trans.level,"s)",
                    ifelse(gene.filters$presence$per.condition, " condition-wise", " across all samples"))
                }
                cat(msg.gene.filters.length,msg.gene.filters.avg.reads,msg.gene.filters.expression.median,msg.gene.filters.expression.mean,
                    msg.gene.filters.expression.quantile,msg.gene.filters.expression.known,msg.gene.filters.expression.custom,msg.gene.filters.biotype,msg.gene.filters.presence,sep=", ")
                cat(". The total number of ",trans.level,"s excluded due to the application of ",trans.level," filters was ",sum(as.numeric(sapply(gene.filter.result,length))),". ",sep="")
                cat("The total (unified) number of ",trans.level,"s  excluded due to the application of all filters was ",length(the.zeros) + length(the.dead),". ",sep="")
              }
              cat("The resulting ",trans.level," counts table was subjected to differential expression analysis for the contrasts ",
                  paste(gsub("_vs_"," versus ",names(html)),collapse=", "),sep="")
              if (length(statistics)>1) {
                cat(" using the Bioconductor packages ",paste(unlist(report.messages$stat[statistics],use.names=FALSE),collapse=", "),". ",sep="")
                if (meta.p!="none") {
                  cat("In order to combine the statistical significance from multiple algorithms and perform meta-analysis, the ")
                  cat(report.messages$meta[[meta.p]],"method was applied. ")
                }
              } else {
                cat(" using the Bioconductor package ",report.messages$stat[[statistics]],". ",sep="")
              }
              if (!is.na(pcut)) if (pcut==1) plasm <- 0.05 else plasm <- pcut
              cat("The per-contrast numbers of differentially expressed ",trans.level,"s were:</p><ul>", sep="")
              msg.contrast <- setNames(character(length(html)), names(html))
              for (cnt in names(html)) {
                if (!is.na(pcut) && length(which(sum.p.list[[cnt]]<plasm))==0) {
                  msg.contrast[cnt] <- paste("<strong>",gsub("_vs_"," versus ",cnt),"</strong>: no differentially expressed ",trans.level,"s were found",
                  " with a p-value threshold of ",plasm,sep="")
                } else if (is.na(pcut)) {
                  msg.contrast[cnt] <- paste("<strong>",gsub("_vs_"," versus ",cnt),"</strong>: no statistical threshold defined", sep="")
                } else {
                  if (adjust.method!="none") {
                    add.text.p <- paste(" (",length(which(custom.p.adjust(sum.p.list[[cnt]],adjust.method)<plasm)),")",sep="")
                    has.fdr.text <- " (FDR or adjusted p-value)"
                  } else
                    add.text.p <- has.fdr.text <- NULL
                  tmp.pval <- sum.p.list[[cnt]]
                  if (length(strsplit(cnt,"_vs_")[[1]])==2) {
                    tmp <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[which(tmp.pval<plasm),,drop=FALSE],log.offset))
                    if (adjust.method!="none") {
                      are.there <- which(custom.p.adjust(tmp.pval,adjust.method)<plasm)
                      if (length(are.there)>0) {
                        if (length(are.there)==1) {
                          tmp.f <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[are.there,,drop=FALSE],log.offset))
                          add.text.fu <- paste0(" (",length(which(tmp.f>=1)),")")
                          add.text.fd <- paste0(" (",length(which(tmp.f<=-1)),")")
                          add.text.fn <- paste0(" (",length(which(abs(tmp.f)<1)),")")
                        } else {
                          tmp.f <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[are.there,,drop=FALSE],log.offset))
                          add.text.fu <- paste0(" (",length(which(tmp.f>=1)),")")
                          add.text.fd <- paste0(" (",length(which(tmp.f<=-1)),")")
                          add.text.fn <- paste0(" (",length(which(abs(tmp.f)<1)),")")
                        }
                      } else add.text.fu <- add.text.fd <- add.text.fn <- NULL
                    } else  add.text.fu <- add.text.fd <- add.text.fn <- NULL
                    msg.contrast[cnt] <- paste0("<strong>", gsub("_vs_"," versus ",cnt), "</strong>: ", length(which(tmp.pval<plasm)), add.text.p,
                                                " statistically significant ",trans.level,"s were found with a p-value ",has.fdr.text," threshold of ",
                                                plasm, " and of these, ",length(which(tmp>=1)),
                                                add.text.fu," were up-regulated, ",length(which(tmp<=-1)), add.text.fd ," were down-regulated and ",
                                                length(which(abs(tmp)<1)),add.text.fn,
                                                " were not differentially expressed according to an absolute fold change cutoff value of 1 in log<sub>2</sub> scale")
                  } else {
                    msg.contrast[cnt] <- paste("<strong>",gsub("_vs_"," versus ",cnt),"</strong>: ",length(which(sum.p.list[[cnt]]<plasm)),add.text.p,
                                               " differentially expressed ",trans.level,"s were found with a p-value",has.fdr.text," threshold of ",
                                               plasm," at least in one condition",sep="")
                  }
                }
              }
              cat(paste("<li>",msg.contrast,"</li>",collapse=""))
              cat("</ul><p>Literature references for all the algorithms used can be found at the end of this report.")
              %></p>
            </div>
            <div class="line"></div>
            <h2 id="summary_input" class="h4"><a href="#summary_input_container" data-toggle="collapse">Input options</a></h2>
            <div id="summary_input_container" class="collapse">
              <strong>Read counts file:</strong> <%=counts.name%><br>
              <strong>Conditions:</strong> <%=paste(names(sample.list),collapse=", ")%><br>
              <strong>Samples included:</strong> <%=paste(unlist(sample.list),collapse=", ")%><br>
              <strong>Samples excluded:</strong><%=ifelse(!is.null(exclude.list) && !is.na(exclude.list), paste(unlist(exclude.list),collapse=", "), "none") %><br>
              <strong>Requested contrasts:</strong> <%=paste(contrast,collapse=", ")%><br>
              <strong>Library sizes:</strong> <% if (!is.null(libsize.list)) {
                cat("<ul>")
                for (n in names(libsize.list)) cat("<li>",paste(n,libsize.list[[n]],sep=": "),"</li>")
                cat("</ul>")
                } else cat("not available","<br>") %>
              <strong>Annotation:</strong> <%=annotation%><br>
              <strong>Organism:</strong> <%=report.messages$org[[org]]%><br>
              <strong>Annotation source:</strong> <%=report.messages$refdb[[refdb]]%><br>
              <strong>Count type:</strong> <%=count.type%><br>
              <% if (count.type=="utr" && utr.flank>0) { cat("<strong>3' UTR flanking :</strong>: ",utr.flank,"<br>") } %>
              <% if (trans.level=="exon") {
                cat("<strong>Exon filters:</strong>")
                if (!is.null(exon.filters)) {
                  cat(paste(names(exon.filters),collapse=", "),"<br>")
                  cat("<ul>")
                  for (ef in names(exon.filters)) {
                    cat("<li><em><span style=\"font-size:1em\">",ef,"</span></em><ul>",sep="")
                    for (efp in names(exon.filters[[ef]])) {
                      if (length(exon.filters[[ef]][[efp]])==1 && is.function(exon.filters[[ef]][[efp]]))
                        cat("<li>custom function</li>")
                      else if (length(exon.filters[[ef]][[efp]])==1)
                        cat("<li>",paste(efp,exon.filters[[ef]][[efp]],sep=": "),"</li>",sep="")
                      else if (length(exon.filters[[ef]][[efp]])>1)
                        cat("<li>",paste(efp,paste(exon.filters[[ef]][[efp]],collapse=", "),sep=": "),"</li>",sep="")
                    }
                    cat("</ul></li>")
                  }
                  cat("</ul>")
                } else cat("none applied","<br>")
              } %>
              <% if (!is.null(preset)) cat("<strong>Analysis preset:</strong>",report.messages$preset[[preset]],"<br>") %>
              <strong>Gene filters:</strong>
              <%
                if (!is.null(gene.filters)) {
                  cat(paste(names(gene.filters),collapse=", "),"<br>")
                  cat("<ul>")
                  for (gf in names(gene.filters)) {
                    if (is.null(gene.filters[[gf]])) next()
                    cat("<li><em><span style=\"font-size:1em\">",gf,"</span></em><ul>",sep="")
                    for (gfp in names(gene.filters[[gf]])) {
                      if (all(is.na(gene.filters[[gf]][[gfp]]))) next()
                      if (length(gene.filters[[gf]][[gfp]])==1 && is.function(gene.filters[[gf]][[gfp]]))
                        cat("<li>custom function</li>")
                      else if (length(gene.filters[[gf]][[gfp]])==1)
                        cat("<li>",paste(gfp,gene.filters[[gf]][[gfp]],sep=": "),"</li>",sep="")
                      else if (length(gene.filters[[gf]][[gfp]])>1)
                        cat("<li>",paste(gfp,paste(gene.filters[[gf]][[gfp]],collapse=", "),sep=": "),"</li>",sep="")
                    }
                    cat("</ul></li>")
                  }
                  cat("</ul>")
                } else cat("none applied","<br>")
              %>
              <strong>Filter application:</strong> <%=report.messages$whenfilter[[when.apply.filter]]%><br>
              <strong>Normalization algorithm:</strong> <%=report.messages$norm[[normalization]]%><br>
              <strong>Normalization arguments:</strong>
              <% if (!is.null(norm.args)) {
                  if (normalization=="each") {
                    for (n in names(norm.args)) {
                      cat("<strong>Statistical arguments for</strong> ",report.messages$norm[[n]],": ",paste(names(norm.args[[n]]),collapse=", "))
                      if (length(norm.args[[n]])>0) {
                        cat("<ul>")
                        for (na in names(norm.args[[n]])) {
                          if (length(norm.args[[n]][[na]])==1 && is.function(norm.args[[n]][[na]]))
                            cat("<li>",na,": ",as.character(substitute(norm.args[[na]])),"</li>",sep="")
                          else if (length(norm.args[[n]][[na]])==1)
                            cat("<li>",paste(na,norm.args[[n]][[na]],sep=": "),"</li>")
                          else if (length(norm.args[[n]][[na]])>1)
                            cat("<li>",paste(na,paste(norm.args[[n]][[na]],collapse=", "),sep=": "),"</li>")
                        }
                        cat("</ul>")
                      } else cat("not available or not required<br>")
                    }
                  } else {
                    cat(paste(names(norm.args),collapse=", "),"<br>")
                    cat("<ul>")
                    for (na in names(norm.args)) {
                      if (all(is.na(norm.args[[na]]))) next()
                      if (length(norm.args[[na]])==1 && is.function(norm.args[[na]]))
                        cat("<li>",as.character(substitute(norm.args[[na]])),"</li>",sep="")
                      else if (length(norm.args[[na]])==1)
                        cat("<li>",paste(na,norm.args[[na]],sep=": "),"</li>")
                      else if (length(norm.args[[na]])>1)
                        cat("<li>",paste(na,paste(norm.args[[na]],collapse=", "),sep=": "),"</li>")
                    }
                    cat("</ul>")
                  }
                } else cat("not available")
              %>
              <strong>Statistical algorithm(s): </strong> <%=paste(unlist(report.messages$stat[statistics],use.names=FALSE),collapse=", ")%><br>
              <%
                if (!is.null(stat.args)) {
                  for (s in names(stat.args)) {
                    cat("<strong>Statistical arguments for ",report.messages$stat[[s]],": </strong>",paste(names(stat.args[[s]]),collapse=", "),sep="")
                    if (length(stat.args[[s]])>0) {
                      cat("<ul>")
                      for (sa in names(stat.args[[s]])) {
                        if (length(stat.args[[s]][[sa]])==1 && is.function(stat.args[[s]][[sa]]))
                          cat("<li>",sa,": ",as.character(substitute(stat.args[[s]][[sa]] )),"</li>",sep="")
                        else if (length(stat.args[[s]][[sa]])==1)
                          cat("<li>",paste(sa,stat.args[[s]][[sa]],sep=": "),"</li>")
                        else if (length(stat.args[[s]][[sa]])>1)
                          cat("<li>",paste(sa,paste(stat.args[[s]][[sa]],collapse=", "),sep=": "),"</li>")
                      }
                      cat("</ul>")
                    } else cat("not available or not required<br>")
                  }
                } else cat("<strong>Statistical arguments not available</strong>")
              %>
              <strong>Meta-analysis method:</strong> <%=report.messages$meta[[meta.p]]%><br>
              <strong>Multiple testing correction: </strong> <%=report.messages$adjust[[tolower(adjust.method)]]%><br>
              <strong>p-value threshold: </strong> <%=if (!is.na(pcut)) cat(pcut) else cat("not available")%><br>
              <strong>Logarithmic tranformation offset: </strong> <%=log.offset%><br>
              <strong>Analysis preset: </strong> <%=if (!is.null(preset)) cat(preset) else ("not available")%><br>
              <strong>Quality control plots: </strong><%=paste(unlist(report.messages$plots[qc.plots],use.names=FALSE),collapse=", ")%><br>
              <strong>Figure format: </strong> <%=paste(fig.format,collapse=", ")%><br>
              <strong>Output directory: </strong> <%=if (!is.na(export.where)) cat(export.where) else cat("default")%><br>
              <strong>Output data: </strong> <%=paste(unlist(report.messages$export[export.what],use.names=FALSE),collapse=", ")%><br>
              <strong>Output scale(s): </strong> <%=paste(unlist(report.messages$export[export.scale],use.names=FALSE),collapse=", ")%><br>
              <strong>Output values: </strong> <%=paste(unlist(report.messages$export[export.values],use.names=FALSE),collapse=", ")%><br>
              <strong>Output statistics: </strong><%=paste(unlist(report.messages$export[export.stats],use.names=FALSE),collapse=", ")%><br>
              <strong>Total run time: </strong><%=exec.time%><br>
            </div>
            <div class="line"></div>
            <h2 id="summary_analysis_filtered" class="h4"><a href="#summary_analysis_filtered_container" data-toggle="collapse"><%=cat("Filtered ",trans.level,"s",sep="")%></a></h2>
            <div id="summary_analysis_filtered_container" class="collapse">
              <strong><%cat("Number of filtered ",trans.level,"s:", sep="")%> </strong><%=length(the.zeros) + length(the.dead)%><em> which is the <strong>union</strong> of</em>
              <ul>
                <li>Filtered because of zero reads: <%=length(the.zeros)%></li>
                <li>
                  Filtered because of exon filters: <%=sum(as.numeric(sapply(exon.filter.result,length)))%>
                  <%
                    if (sum(as.numeric(sapply(exon.filter.result,length)))!=0) {
                      cat("<em> which is the <strong>union</strong> of</em>")
                      cat("<ul>")
                      for (n in names(exon.filter.result)) {
                        cat("<li>",n,": ",length(exon.filter.result[[n]]),"</li>")
                      }
                      cat("</ul>")
                    }
                  %>
                </li>
                <li>
                  Filtered because of gene filters: <%=length(unique(unlist(gene.filter.result)))%><em> which is the <strong>union</strong> of</em>
                  <ul>
                    <%
                      for (n in names(gene.filter.result)) {
                        if (!is.list(gene.filter.result[[n]])) {
                          if (!is.null(gene.filter.result[[n]]))
                            cat("<li>",n,": ",length(unlist(gene.filter.result[[n]])),
                                " genes with filter cutoff value ",gene.filter.cutoff[[n]],"</li>",sep="")
                        } else {
                          cat("<li>",n,": ",length(unlist(gene.filter.result[[n]])),
                              " ",trans.level,"s further decomposed to (filter name, filtered ",trans.level,"s, filter cutoff):",sep="")
                          cat("<ul>")
                          for (nn in names(gene.filter.result[[n]])) {
                            if (!is.null(gene.filter.result[[n]][[nn]]))
                              cat("<li>",nn,": ",length(unlist(gene.filter.result[[n]][[nn]])),
                                  " ",trans.level,"s with filter cutoff value ",paste(gene.filter.cutoff[[n]][[nn]],
                                  collapse=", "),"</li>",sep="")
                          }
                          cat("</ul></li>")
                        }
                      }
                    %>
                  </ul>
                </li>
              </ul>
            </div>
            <div class="line"></div>
            <h2 id="summary_analysis_nde" class="h4"><a href="#summary_analysis_nde_container" data-toggle="collapse"><%cat("Differentially expressed ",trans.level,"s",sep="")%></a></h2>
            <div id="summary_analysis_nde_container" class="collapse">
              <strong><%cat("Number of differentially expressed ",trans.level,"s per contrast: ",sep="")%></strong>
              <%
                if (!is.na(pcut) && pcut==1)
                  cat("The p-value cutoff during the analysis was set to 1 so as to retrieve the total ",trans.level," list which passed the ",
                      "filtering procedure. Each ",trans.level," in the list is accompanied by its statistical scores (p-value, FDR, etc.)<br>")
              %>
              <ul><%
                for (cnt in names(html)) {
                  cat("<li><strong>",cnt,": </strong>",sep="")
                  if (!is.na(pcut) && length(which(sum.p.list[[cnt]]<pcut))==0) {
                    cat("no differentially expressed ",trans.level,"s</li>",sep="")
                  } else if (is.na(pcut)) {
                    cat("no statistical threshold defined</li>")
                  } else {
                    if (!is.na(pcut) && pcut==1) {
                      plasm <- 0.05
                    } else {
                      plasm <- pcut
                    }
                    if (adjust.method!="none") {
                      add.text.p <- paste("(",length(which(custom.p.adjust(sum.p.list[[cnt]],adjust.method)<plasm)),")",sep="")
                      has.fdr.text <- "(FDR or adjusted p-value)"
                    } else add.text.p <- has.fdr.text <- NULL
                    if (length(strsplit(cnt,"_vs_")[[1]])==2) {
                      tmp.p <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[which(sum.p.list[[cnt]]<plasm),,drop=FALSE],log.offset))
                      if (adjust.method!="none") {
                        are.there <- which(custom.p.adjust(sum.p.list[[cnt]],adjust.method)<plasm)
                        if (length(are.there)>0) {
                          if (length(are.there)==1) {
                            tmp.f <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[are.there,,drop=FALSE],log.offset))
                            add.text.fu <- paste(" (",length(which(tmp.f>=1)),")",sep="")
                            add.text.fd <- paste(" (",length(which(tmp.f<=-1)),")",sep="")
                            add.text.fn <- paste(" (",length(which(abs(tmp.f)<1)),")",sep="")
                          } else {
                            tmp.f <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[are.there,,drop=FALSE],log.offset))
                            add.text.fu <- paste(" (",length(which(tmp.f>=1)),")",sep="")
                            add.text.fd <- paste(" (",length(which(tmp.f<=-1)),")",sep="")
                            add.text.fn <- paste(" (",length(which(abs(tmp.f)<1)),")",sep="")
                          }
                        } else add.text.fu <- add.text.fd <- add.text.fn <- NULL
                      } else add.text.fu <- add.text.fd <- add.text.fn <- NULL
                      cat(length(which(sum.p.list[[cnt]]<plasm))," ",add.text.p," statistically significant ",trans.level,"s of which ",length(which(tmp.p>=1)),
                          add.text.fu," up regulated, ",length(which(tmp.p<=-1)),add.text.fd," down regulated and ",length(which(abs(tmp.p)<1)),
                          add.text.fn," not differentially expressed according to a p-value ",has.fdr.text," threshold of ",plasm," and an absolute ",
                          " fold change cutoff value of 1 in log<sub>2</sub> scale.",sep="")
                    } else
                      cat(length(which(sum.p.list[[cnt]]<plasm))," ",add.text.p," statistically significant, differentially expressed in at",
                          " least one condition at a p-value ",has.fdr.text," threshold of ",plasm,".",sep="")
                    if (length(statistics)>1 && meta.p!="none") {
                      cat(" These numbers refer to the combined analysis performed by metaseqR.",
                          " Per statistical algorithm, the differentially expressed ",trans.level,"s are:", sep="")
                      cat("<ul>")
                      for (s in statistics) {
                        cat("<li><strong>",report.messages$stat[[s]],": </strong>",sep="")
                        if (adjust.method!="none") {
                          add.text.p <- paste("(",length(which(custom.p.adjust(cp.list[[cnt]][,s],adjust.method)<plasm)),")",sep="")
                          has.fdr.text <- "(FDR or adjusted p-value)"
                        } else add.text.p <- has.fdr.text <- NULL
                        if (length(strsplit(cnt,"_vs_")[[1]])==2) {
                          tmp.p <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[which(cp.list[[cnt]][,s]<plasm),,drop=FALSE],log.offset))
                          if (adjust.method!="none") {
                            tmp.f <- log2(make.fold.change(cnt,sample.list,norm.genes.expr[which(custom.p.adjust(cp.list[[cnt]][,s],adjust.method)<plasm),,drop=FALSE],log.offset))
                            add.text.fu <- paste("(",length(which(tmp.f>=1)),")",sep="")
                            add.text.fd <- paste("(",length(which(tmp.f<=-1)),")",sep="")
                            add.text.fn <- paste("(",length(which(abs(tmp.f)<1)),")",sep="")
                          } else add.text.fu <- add.text.fd <- add.text.fn <- NULL
                          cat(length(which(cp.list[[cnt]][,s]<plasm))," ",add.text.p," statistically significant ",trans.level,"s of which ",length(which(tmp.p>=1))," ",
                              add.text.fu," up regulated, ",length(which(tmp.p<=-1)),add.text.fd," down regulated and ",length(which(abs(tmp.p)<1))," ",
                              add.text.fn," not differentially expressed according to a p-value "," ", has.fdr.text," threshold of ",plasm," and an absolute ",
                              "fold change cutoff value of 1 in log<sub>2</sub> scale.",sep="")
                        } else {
                          cat(length(which(cp.list[[cnt]][,s]<plasm))," ",add.text.p," statistically significant, differentially expressed in at",
                              " least one condition at a p-value ",has.fdr.text," threshold of ",plasm,".",sep="")
                        }
                        cat("</li>")
                      }
                      cat("</ul>")
                    }
                  }
                }
              %></ul>
            </div>
            <%
              if (run.log) {
                cat('<div class="line"></div>')
                cat('<h2 id="summary_analysis_log" class="h4"><a href="#summary_analysis_log_container" data-toggle="collapse">Run log</a></h2>')
                cat('<div id="summary_analysis_log_container" class="collapse"><pre class="pre-scrollable"><code>')
                log.string <- paste(readLines(file.path(PROJECT.PATH$logs,"metaseqr_run.log")),collapse="---EOL---")
                cat(gsub("---EOL---","<br>",log.string))
                cat('</code></pre></div>')
              }
            %>
            <div class="line"></div>
          </div>
        </div>
        <div class="row">
          <div class="col">
            <h1 id="figures" class="section-header"><a href="#figures_container" data-toggle="collapse">Figures</a></h1>
          </div>
        </div>
        <div id="figures_container" class="row collapse m-2">
          <div class="col">
            <%
            if (!is.null(fig.raw$mds)) {
              subsection.one.image('figures_mds', 'Multidimensional scaling', fig.raw$mds, report.messages$explain$mds, "Multidimensional scaling", "MDS")
            }
            if (!is.null(fig.raw$biodetection)) {
              subsection.one.gallery('figures_biodetection', 'Biotype detection', fig.raw$biodetection, report.messages$explain$biodetection, "Biotype detection", "biotype detection")
            }
            if (!is.null(fig.raw$countsbio)) {
              subsection.one.gallery('figures_countsbio', 'Biotype detection counts', fig.raw$countsbio, report.messages$explain$countsbio, "Biotype detection counts", "biotype detection counts")
            }
            if (!is.null(fig.raw$saturation)) {
              subsection('figures_saturation', 'Read and biotype saturation', function() {
                gid  <- gallery_id()
                path <- gsub(PROJECT.PATH$main,".",fig.raw$saturation$biotype)
                cat('<p class="text-justify">',report.messages$explain$saturation,'</p>', sep="")
                cat('<div class="text-center"><figure class="figure">',
                    '<a href="',path,'" data-fancybox="',gid,'" data-caption="Read saturation per biotype for all samples">',
                    '<img src="',path,'" alt="sample biotype saturation" class="img-fluid img-thumbnail"></a>',
                    '<figcaption class="figure-caption">Read saturation per biotype for all samples</figcaption></figure></div>',sep="")
                path <- gsub(PROJECT.PATH$main,".",fig.raw$saturation$sample)
                cat('<h5 class="text-center">Read saturation per sample for all biotypes</h5>')
                cat('<div class="d-flex align-items-center align-content-around flex-wrap">')
                for (s in path) {
                  cat('<div class="text-center"><a href="',s,'" data-fancybox="',gid,'" data-caption="Biotype saturation per sample">',
                      '<img src="',s,'" alt="sample biotype saturation" class="img-fluid img-thumbnail"></a></div>', sep="")
                }
                cat('</div>')
              });
            }
            if (!is.null(fig.raw$readnoise)) {
              subsection.one.image('figures_readnoise', 'RNA-Seq reads noise', fig.raw$readnoise, report.messages$explain$readnoise, "Read noise")
            }
            if (!is.null(fig.raw$correl)) {
              subsection.multi.image('figures_correl', 'Correlation plots', c(fig.raw$correl$heatmap, fig.raw$correl$correlogram),
                                    report.messages$explain$correl, c('Correlation heatmap', 'Data correlogram'))
            }
            if (!is.null(fig.raw$pairwise)) {
              subsection.one.image('figures_pairwise', 'Pairwise scatterplots', fig.raw$pairwise, report.messages$explain$pairwise, "Pairwise sample scatterplots")
            }
            if (!is.null(fig.unorm$boxplot) || !is.null(fig.norm$boxplot)) {
              figures <- c()
              captions <- c()
              if (!is.null(fig.unorm$boxplot)) {
                figures <- c(figures, fig.unorm$boxplot)
                captions <- c(captions, 'Boxplot of un-normalized data')
              }
              if (!is.null(fig.norm$boxplot)) {
                figures <- c(figures, fig.norm$boxplot)
                captions <- c(captions, 'Boxplot of normalized data')
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_boxplot', 'Boxplots', figures, report.messages$explain$boxplot, captions)
              }
            }
            if (!is.null(fig.unorm$rnacomp) || !is.null(fig.norm$rnacomp)) {
              figures <- c()
              captions <- c()
              if (!is.null(fig.unorm$rnacomp)) {
                figures <- c(figures, fig.unorm$rnacomp)
                captions <- c(captions, 'RNA composition of un-normalized data')
              }
              if (!is.null(fig.unorm$rnacomp)) {
                figures <- c(figures, fig.norm$rnacomp)
                captions <- c(captions, 'RNA composition of normalized data')
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_rnacomp', 'RNA composition', figures, report.messages$explain$rnacomp, captions)
              }
            }
            if (!is.null(fig.unorm$gcbias) || !is.null(fig.norm$gcbias)) {
              figures <- c()
              captions <- c()
              if (!is.null(fig.unorm$gcbias)) {
                figures <- c(figures, fig.unorm$gcbias)
                captions <- c(captions, 'GC content bias un-normalized')
              }
              if (!is.null(fig.unorm$gcbias)) {
                figures <- c(figures, fig.norm$gcbias)
                captions <- c(captions, 'GC content bias normalized')
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_gcbias', 'GC content bias', figures, report.messages$explain$gcbias, captions)
              }
            }
            if (!is.null(fig.unorm$lengthbias) || !is.null(fig.norm$lengthbias)) {
              figures <- c()
              captions <- c()
              if (!is.null(fig.unorm$lengthbias)) {
                figures <- c(figures, fig.unorm$lengthbias)
                captions <- c(captions, 'Gene/transcript length bias un-normalized')
              }
              if (!is.null(fig.unorm$lengthbias)) {
                figures <- c(figures, fig.norm$lengthbias)
                captions <- c(captions, 'Gene/transcript length bias normalized')
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_lengthbias', 'Gene/transcript length bias', figures, report.messages$explain$lengthbias, captions)
              }
            }
            if (!is.null(fig.unorm$meandiff) || !is.null(fig.norm$meandiff)) {
              figures  <- c()
              captions <- c()
              groups   <- c()
              if (is.null(fig.unorm$meandiff))
                nn <- names(fig.norm$meandiff)
              else
                nn <- names(fig.unorm$meandiff)
              for (n in nn) {
                g <- paste0("Mean-difference plots for the replicates of ",n)
                if (is.null(fig.unorm$meandiff[[n]])) {
                  l <- length(fig.norm$meandiff[[n]])
                } else {
                  l <- length(fig.unorm$meandiff[[n]])
                }
                for (i in 1:l) {
                  if (!is.null(fig.unorm$meandiff[[n]])) {
                    figures  <- c(figures, fig.unorm$meandiff[[n]][i])
                    captions <- c(captions, 'Un-normalized mean-difference plots')
                    groups   <- c(groups, g)
                  }
                  if (!is.null(fig.norm$meandiff[[n]])) {
                    figures  <- c(figures, fig.norm$meandiff[[n]][i])
                    captions <- c(captions, 'Normalized mean-difference plots')
                    groups   <- c(groups, g)
                  }
                }
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_meandiff', 'Mean-difference', figures, report.messages$explain$meandiff, captions, groups)
              }
            }
            if (!is.null(fig.unorm$meanvar) || !is.null(fig.norm$meanvar)) {
              figures  <- c()
              captions <- c()
              groups   <- c()
              if (is.null(fig.unorm$meanvar))
                nn <- names(fig.norm$meanvar)
              else
                nn <- names(fig.unorm$meanvar)
              for (n in nn) {
                g <- paste0("Mean-variance plots for the replicates of ",n)
                if (is.null(fig.unorm$meanvar[[n]])) {
                  l <- length(fig.norm$meanvar[[n]])
                } else {
                  l <- length(fig.unorm$meanvar[[n]])
                }
                for (i in 1:l) {
                  if (!is.null(fig.unorm$meanvar[[n]])) {
                    figures  <- c(figures, fig.unorm$meanvar[[n]][i])
                    captions <- c(captions, 'Un-normalized mean-variane plots')
                    groups   <- c(groups, g)
                  }
                  if (!is.null(fig.norm$meanvar[[n]])) {
                    figures  <- c(figures, fig.norm$meanvar[[n]][i])
                    captions <- c(captions, 'Normalized mean-variance plots')
                    groups   <- c(groups, g)
                  }
                }
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_meanvar', 'Mean-variance', figures, report.messages$explain$meanvar, captions, groups)
              }
            }
            if (!is.null(fig.other$filtered)) {
              subsection.one.image('figures_filtered', paste0("Chromosome and biotype distribution of filtered ",trans.level,"s"),
                                    fig.other$filtered, report.messages$explain$filtered, paste0("Filtered ",trans.level,"s"))
            }
            if (!is.null(fig.stat$volcano)) {
              subsection('figures_volcano', 'Volcano plots', function () {
                cat('<p class="text-justify">',report.messages$explain$volcano,'</p>', sep="")
                cat('<div class="d-flex m-2 p-2 justify-content-center align-content-around align-items-center flex-wrap">')
                nn <- names(contrast.list)
                counter <- 1
                for (n in nn) {
                  cat('<div class="p-2"><a href="Javascript:;" id="volcano_',counter,'_button" class="btn btn-outline-dark">',
                    'Volcano plot for the contrast ',parse.name(n),'</a></div>',sep="")
                    counter <- counter + 1
                }
                cat('</div>')
              })
            }
            if (!is.null(fig.stat$deheatmap)) {
              figures  <- c()
              captions <- c()
              nn <- names(fig.stat$deheatmap)
              counter <- 1
              for (n in nn) {
                figures  <- c(figures, fig.stat$deheatmap[n])
                captions <- c(captions, paste0("DEG heatmap for the contrast ",parse.name(n)))
                counter <- counter + 1
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_deheatmap', 'DEG heatmaps', figures, report.messages$explain$deheatmap, captions)
              }
            }
            if (!is.null(fig.stat$biodist)) {
              figures  <- c()
              captions <- c()
              nn <- names(fig.stat$biodist)
              counter <- 1
              for (n in nn) {
                figures  <- c(figures, fig.stat$biodist[n])
                captions <- c(captions, paste0("Biotype distribution in DEG for the contrast ",parse.name(n)))
                counter <- counter + 1
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_biodist', 'Chromosome and biotype distributions of DEGs', figures, report.messages$explain$biodist, captions)
              }
            }
            if (!is.null(fig.venn$venn)) {
              figures  <- c()
              captions <- c()
              alts     <- c()
              nn <- names(fig.venn$venn)
              counter <- 1
              for (n in nn) {
                figures  <- c(figures, fig.venn$venn[n])
                captions <- c(captions, paste0("Venn diagram for the contrast ",parse.name(n),'<br><a href="lists/venn_categories_',n,'.txt" download>Download lists</a>'))
                alts     <- c(alts, paste0("Venn diagram for the contrast ",parse.name(n)))
                counter <- counter + 1
              }
              if (length(figures) > 0) {
                subsection.multi.image('figures_venn', 'Meta-analysis Venn diagrams', figures, report.messages$explain$venn, captions, NULL, alts, alts)
              }
            }
            %>
            <div class="p-t-2 p-b-2">
              <p class="h6">
                Get all the figures in
                <%=paste(paste("<a href=\"plots/metaseqr_figures_",fig.format,".zip\" class=\"btn btn-outline-dark\" download>",fig.format,"</a>",sep=""),collapse=", ")%>
                format.
              </p>
            </div>
            <div class="line"></div>
          </div>
        </div>
        <div class="row">
          <div class="col">
            <h1 id="results" class="section-header"><a href="#results_container" data-toggle="collapse">Results</a></h1>
          </div>
        </div>
        <div id="results_container" class="row collapse m-2">
          <div class="col">
          <%
            if (!is.null(html)) {
              nn      <- names(html)
              counter <- 1
              for (n in nn) {
                if (is.null(report.top))
                  top.text <- paste("all the statistically significant ",	trans.level,"s",sep="")
                else
                  top.text <- paste("the top ",round(100*report.top),"% statistically significant ",trans.level,"s (use the download links below the table to retrieve the whole list)",sep="")
                subsection(paste0("table_container_", counter), paste0('DEG table for the contrast ', parse.name(n)), function() {
                  cat('<p class="text-justify">The following table presents ',top.text,' for the contrast <strong>',parse.name(n),
                      '</strong>. The fields of the table correspond to the requested features to be exported. The table can be ',
                      'searched using the search field on the top right.</p>')
                  cat('<div class="">', html[[n]], '</div>', sep="")
                  cat("<p class=\"mt-2\"><strong><a href=\"lists/metaseqr_sig_out_",n,".txt.gz\" download>Download</a> the DEG result list for ",parse.name(n),".</strong></p>",sep="")
                  if (!is.null(gene.counts.zero) || !is.null(gene.counts.dead)) {
                      cat("<p><strong><a href=\"lists/metaseqr_all_out_",n,".txt.gz\" download>Download</a> the whole result list for ",parse.name(n),".</strong></p>",sep="")
                  }
                })
                counter <- counter + 1
              }
              if (export.counts.table) {
                cat('<div class="p-t-2 p-b-2">')
                cat('<p class="h6">')
                if (file.exists(file.path(PROJECT.PATH[["lists"]],"raw_counts_table.txt.gz")))
                  cat("<a href=\"lists/raw_counts_table.txt.gz\" class=\"btn btn-outline-dark\" download>Download</a> the raw read counts table for the experiment.</p><p class=\"h6\">",sep="")
                cat("<a href=\"lists/normalized_counts_table.txt.gz\" class=\"btn btn-outline-dark\" download>Download</a> the normalized read counts table for the experiment.</p>",sep="")
                cat('</div>')
                cat('<div class="line"></div>')
              }
            }
          %>
          </div>
        </div>
        <div class="row">
          <div class="col">
            <h1 id="references" class="section-header"><a href="#references_container" data-toggle="collapse">References</a></h1>
          </div>
        </div>
        <div id="references_container" class="row collapse m-2">
          <div class="col">
            <%
              refs <- unique(c(
                report.messages$references$filein[[file.type]],
                report.messages$references$norm[[normalization]],
                unlist(report.messages$references$stat[statistics],use.names=FALSE),
                unlist(report.messages$references$figure[qc.plots],use.names=FALSE),
                report.messages$references$multiple[[adjust.method]],
                report.messages$references$meta[[meta.p]],
                "Moulos, P. and Hatzis P. (2014). Systematic integration of RNA-Seq statistical algorithms for accurate detection of differential gene expression patterns. Nucleic Acids Research, 10.1093/nar/gku1273."
              ))
              cat("<ol>")
              cat(paste("<li>",refs, "</li>", sep="", collapse=""))
              cat("</ol>")
            %>
          </div>
        </div>
        <footer class="small">
          Copyright &copy; 2013 <a href="http://www.fleming.gr/research/molecular-biology/talianidis-lab" target="_blank">Talianidis Lab</a>/<a href="http://www.fleming.gr/" target="_blank">BSRC Alexander Fleming</a>,
          Original design by <a href="http://epigenomics.fleming.gr/pmoulos/" target="_blank">Panagiotis Moulos</a> - Custom design for RNAdetector by <a href="http://www.alaimos.com/" target="_blank">Salvatore Alaimo</a>
        </footer>
      </div>
    </div>

    <script src="media/scripts.js"></script>
    <script>
      $(document).ready(function() {
        $('#sidebarCollapse').on('click', function() {
          $('#sidebar').toggleClass('active');
          $(this).toggleClass('active');
        });
        var h = function() {
          if (window.location.hash) {
            $('#sidebar-menu')
              .find('li.active')
              .removeClass('active');
            $('#sidebar-menu')
              .find('a[href="' + window.location.hash + '"]')
              .parent()
              .addClass('active');
            $('#sidebar-menu')
              .find('li.active')
              .closest('ul.collapse')
              .collapse('show');
            var link = $('#sidebar-menu').find('a[href="' + window.location.hash + '"]').attr('href');
            var container = $(link+'_container');
            if (!container.hasClass('show')) {
              container.collapse('toggle');
            }
            $('html, body').animate({
              scrollTop: $(link).offset().top
            }, 2000);
          }
        };
        $(window).on('hashchange', function(e) {
          e.preventDefault();
          h();
        });
        h();
      });
    </script>
    <script>
      $(function() {
        var initVolcano = function(counter, data) {
          $('#volcano_' + counter + '_button').click(function () {
            var code = '<div class="volcano_external"><div class="volcano_container"><div class="volcano" id="volcano_' + counter + '"></div></div></div>';
            $.fancybox.open(code, {
              closeExisting: true,
              afterShow: function() {
                $('#volcano_' + counter).highcharts(data);
              }
            });
          });
        };
        <%
          if (!is.null(fig.stat$volcano)) {
            nn <- names(contrast.list)
            counter <- 1
            for (n in nn) {
              fc <- log2(make.fold.change(n,sample.list,norm.genes.expr,1))
              for (contrast in colnames(fc)) {
                json <- diagplot.volcano(fc[,contrast],sum.p.list[[n]],contrast,alt.names=gene.data.expr$gene_name,output="json")
                cat('initVolcano(', counter,',',json,');', sep="")
                counter <- counter+1
              }
            }
          }
        %>
      });
    </script>
    <script>
      $(function() {
        var container = $('#results_container');
        var s = $('#sidebar');
        var tableHandler = function(elem, first) {
          var _init = false;
          var _tbl = undefined;
          var _table_container = undefined;
          var _external_container = undefined;
          var _resize = function () {
            if (!container.hasClass('show')) return;
            var e = $(elem);
            if (!_init) {
              _external_container = e.parent().parent().parent().parent();
              _table_container = e.parent();
            }
            if (!_external_container.parent().hasClass('show')) return;
            _table_container.removeAttr('style');
            var sw = (!_init || (s.outerWidth(true) - s.innerWidth()) < 0) ? 0 : s.width();
            var w = _external_container.width() - sw;
            _table_container.css('max-width', w + 'px')
            if (!_tbl) {
              _tbl = e.DataTable({
                scrollX: true,
                paging: true
              });
            } else {
              _tbl.columns.adjust();
            }
            _init = true;
          };
          $(window).resize(function() {
            _resize();
          });
          return _resize;
        };
        var tables = $('.datatable').toArray();
        var handlers = new Array(tables.length);
        for (var i = 0; i < tables.length; i++) {
          handlers[i] = tableHandler(tables[i], i == 0);
          handlers[i]();
        }
        container.on('shown.bs.collapse', function (e) {
          for (var i = 0; i < tables.length; i++) {
            handlers[i]();
          }
        });
      });
    </script>
  </body>
</html>
